blockdevice/stats.go

```diff
@@ -0,0 +1,182 @@
+	ReadSectors uint64
+	// ReadTicks is the total number of milliseconds spent by all reads.
+	ReadTicks uint64
+	// WriteIOs is the total number of writes completed successfully.
+	WriteIOs uint64
+	// WriteMerges is the number of reads merged.
+	WriteMerges uint64
+	// WriteSectors is the total number of sectors written successfully.
+	WriteSectors uint64
+	// WriteTicks is the total number of milliseconds spent by all writes.
+	WriteTicks uint64
+	// IOsInProgress is number of I/Os currently in progress.
+	IOsInProgress uint64
+	// IOsTotalTicks is the number of milliseconds spent doing I/Os.
+	// This field increases so long as IosInProgress is nonzero.
+	IOsTotalTicks uint64
+	// WeightedIOTicks is the weighted number of milliseconds spent doing I/Os.
+	// This can also be used to estimate average queue wait time for requests.
+	WeightedIOTicks uint64
+	// DiscardIOs is the total number of discards completed successfully.
+	DiscardIOs uint64
+	// DiscardMerges is the number of discards merged.
+	DiscardMerges uint64
+	// DiscardSectors is the total number of sectors discarded successfully.
+	DiscardSectors uint64
+	// DiscardTicks is the total number of milliseconds spent by all discards.
+	DiscardTicks uint64
+}
+
+// Diskstats combines the device Info and IOStats
+type Diskstats struct {
+	Info
+	IoStats
+	// IoStatsCount contains the number of io stats read.  For kernel versions
+	// 4.18+, there should be 18 fields read.  For earlier kernel versions this
+	// will be 14 because the discard values are not available.
+	IoStatsCount int
+}
+
+const (
+	procDiskstatsPath   = "diskstats"
+	procDiskstatsFormat = "%d %d %s %d %d %d %d %d %d %d %d %d %d %d %d %d %d %d"
+	sysBlockPath        = "block"
+	sysBlockStatFormat  = "%d %d %d %d %d %d %d %d %d %d %d %d %d %d %d"
+)
+
+// ReadProcDiskstats reads the diskstats file and returns
+// an array of Diskstats (one per line/device)
+func ReadProcDiskstats(procfs string) ([]Diskstats, error) {
+	file, err := os.Open(path.Join(procfs, procDiskstatsPath))
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	diskstats := []Diskstats{}
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		d := &Diskstats{}
+		d.IoStatsCount, err = fmt.Sscanf(scanner.Text(), procDiskstatsFormat,
+			&d.MajorNumber,
+			&d.MinorNumber,
+			&d.DeviceName,
+			&d.ReadIOs,
+			&d.ReadMerges,
+			&d.ReadSectors,
+			&d.ReadTicks,
+			&d.WriteIOs,
+			&d.WriteMerges,
+			&d.WriteSectors,
+			&d.WriteTicks,
+			&d.IOsInProgress,
+			&d.IOsTotalTicks,
+			&d.WeightedIOTicks,
+			&d.DiscardIOs,
+			&d.DiscardMerges,
+			&d.DiscardSectors,
+			&d.DiscardTicks,
+		)
+		// The io.EOF error can be safely ignored because it just means we read fewer than
+		// the full 18 fields.
+		if err != nil && err != io.EOF {
+			return diskstats, err
+		}
+		if d.IoStatsCount == 14 || d.IoStatsCount == 18 {
+			diskstats = append(diskstats, *d)
+		}
+	}
+	return diskstats, nil
```

Comment:
The scanner's `Err()` function needs to be considered and returned here.

Reply:
Good catch, I'll fix this.